### PR TITLE
Add liquidity source breakdown to asset-swapper

### PR DESCRIPTION
**Pull request by dave4506**
## Description

Adds an addition `sourceBreakdown` entry to `SwapQuote` provides a breakdown of the orders by liquidity source.

Example:
```
{
    "Kyber": 0.2,
    "Native": 0.4,
    "Uniswap": 0.1,
    "Eth2Dai": 0.3
}
```

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 1]
```diff
@@ -266,6 +279,49 @@ export class SwapQuoteCalculator {
...
+        return Object.keys(aggregatedFillAmountBySource).reduce((acc: SwapQuoteOrdersBreakdown, source: string): SwapQuoteOrdersBreakdown => {
+            return {
+                ...{
+                    [source]: aggregatedFillAmountBySource[source].dividedBy(totalAssetAmount),
+                },
+                ...acc,
+            };
+        }, breakdown);
+    }
 }
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts]: There's actually a more direct way to get this info.  `getMarketSellOrdersAsync()` now returns a `OptimizedMarketOrder[]` type, which each have a `fill` field that specifies the source and asset amount.

Comment (dave4506) [reply packages/asset-swapper/src/utils/swap_quote_calculator.ts]: <face palm>

File: packages/asset-swapper/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@0x/asset-swapper",
-    "version": "4.0.1",
+    "version": "4.1.0",
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/package.json]: Eh? Are we publishing this separately or something?

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 3]
```diff
@@ -398,6 +403,31 @@ export class SwapQuoteCalculator {
...
+        orders: OptimizedMarketOrder[],
+        operation: MarketOperation,
+    ): SwapQuoteOrdersBreakdown {
+        // HACK: to shut up linter
+        const breakdown: SwapQuoteOrdersBreakdown = {};
+
+        // total asset amount (accounting for slippage protection)
+        const totalAssetAmount = orders.reduce((amount: BigNumber, order: OptimizedMarketOrder): BigNumber => {
+            return amount.plus(operation === MarketOperation.Buy ? order.makerAssetAmount : order.takerAssetAmount);
+        }, constants.ZERO_AMOUNT);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts]: Nit
```ts
const totalAssetAmount = BigNumber.sum(
    ...[0, orders.map(o => operation === MarketOperation.Buy ? o.makerAssetAmount : o.takerAssetAmount)],
);
```

File: packages/asset-swapper/package.json (hunk 2 above)
Comment (dave4506) [reply packages/asset-swapper/package.json]: Ope, I was changing that when debugging something, my bad

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts (hunk 3 above)
Comment (dave4506) [reply packages/asset-swapper/src/utils/swap_quote_calculator.ts]: clean af

**Comment by fragosti** [comment]
Some of the other dex aggregators have a return format of:
```
[
{
  name: 'Kyber',
  amount: ...
  proprotion: ...
}
]
```
I feel like this is more extensible which is what we want in the API. Can we change the format to be more like this?

**Pull request merged**